File: day2/puzzle2.py
#!/usr/bin/env python3

import re

data_file_name = "puzzle2_data.txt"

with open(data_file_name, 'r') as file:
    directions = [line for line in file]

aim_count = 0
horizontal_count = 0
depth_count = 0

for direction in directions:
    magnitude = int(re.search(r'\d+', direction).group())

    if "down" in direction:
        aim_count += magnitude
    elif "up" in direction:
        aim_count -= magnitude
    elif "forward" in direction:
        horizontal_count += magnitude
        depth_count += (aim_count * magnitude)
        

print(f"Aim: {aim_count}, horizontal: {horizontal_count}, depth: {depth_count} = {horizontal_count}*{depth_count} = {horizontal_count * depth_count}")

File: day2/swift/puzzle2.swift
#!/usr/bin/swift

import Foundation

let dataFilePath = "../day2_data.txt"

// Read file
let fileContents = try String(contentsOfFile: dataFilePath, encoding: String.Encoding.utf8)
let directions = fileContents.split(whereSeparator: \.isNewline).map( { $0 } ) // Force unwrapping not ideal

// Variables to store counts in
var aimCount = 0
var horizontalCount = 0
var depthCount = 0

// Update counter for each direction instruction
directions.forEach {
    let magnitude = Int($0.filter("0123456789".contains))!
    switch $0 {
        case _ where $0.contains("down"): aimCount += magnitude
        case _ where $0.contains("up"): aimCount -= magnitude
        case _ where $0.contains("forward"):
            horizontalCount += magnitude
            depthCount += (aimCount * magnitude)
        default: break
    }
}

print("Aim: \(aimCount), horizontal: \(horizontalCount), depth: \(depthCount) = \(horizontalCount)*\(depthCount) = \(horizontalCount * depthCount)")